File: internal/model/channel.go
package model

import (
	"encoding/json"
	"strconv"
	"strings"

	"github.com/xuanli27/octopus/internal/transformer/outbound"
)

type AutoGroupType int

const (
	AutoGroupTypeNone  AutoGroupType = 0 //不自动分组
	AutoGroupTypeFuzzy AutoGroupType = 1 //模糊匹配
	AutoGroupTypeExact AutoGroupType = 2 //准确匹配
	AutoGroupTypeRegex AutoGroupType = 3 //正则匹配
)

func (t AutoGroupType) Valid() bool {
	switch t {
	case AutoGroupTypeNone, AutoGroupTypeFuzzy, AutoGroupTypeExact, AutoGroupTypeRegex:
		return true
	default:
		return false
	}
}

func ParseAutoGroupSettingValue(value string) (AutoGroupType, bool) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "", "false":
		return AutoGroupTypeNone, true
	case "true":
		return AutoGroupTypeFuzzy, true
	}
	parsed, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		return AutoGroupTypeNone, false
	}
	mode := AutoGroupType(parsed)
	return mode, mode.Valid()
}

type ChannelWSMode string

const (
	ChannelWSModeInherit     ChannelWSMode = "inherit"
	ChannelWSModeOff         ChannelWSMode = "off"
	ChannelWSModePassthrough ChannelWSMode = "passthrough"
	ChannelWSModeTransform   ChannelWSMode = "transform"
)

func (m ChannelWSMode) Normalize() ChannelWSMode {
	switch m {
	case ChannelWSModeOff, ChannelWSModePassthrough, ChannelWSModeTransform:
		return m
	default:
		return ChannelWSModeInherit
	}
}

type Channel struct {
	ID            int                   `json:"id" gorm:"primaryKey"`
	Name          string                `json:"name" gorm:"unique;not null"`
	Type          outbound.OutboundType `json:"type"`
	Enabled       bool                  `json:"enabled" gorm:"default:true"`
	BaseUrls      []BaseUrl             `json:"base_urls" gorm:"serializer:json"`
	Keys          []ChannelKey          `json:"keys" gorm:"foreignKey:ChannelID"`
	Model         string                `json:"model"`
	CustomModel   string                `json:"custom_model"`
	ProxyMode     ProxyUsageMode        `json:"proxy_mode" gorm:"type:varchar(16);not null;default:'direct'"`
	ProxyConfigID *int                  `json:"proxy_config_id"`
	Proxy         bool                  `json:"-" gorm:"default:false"`
	AutoSync         bool          `json:"auto_sync" gorm:"default:false"`
	SkipHealthProbe  bool          `json:"skip_health_probe" gorm:"default:false"` // issue #102: skip group health probes
	AutoGroup        AutoGroupType `json:"auto_group" gorm:"default:0"`
	CustomHeader  []CustomHeader        `json:"custom_header" gorm:"serializer:json"`
	WSMode        ChannelWSMode         `json:"ws_mode" gorm:"type:varchar(16);not null;default:'inherit'"`
	ParamOverride *string               `json:"param_override"`
	ChannelProxy  *string               `json:"-" gorm:"column:channel_proxy"`
	Stats         *StatsChannel         `json:"stats,omitempty" gorm:"foreignKey:ChannelID"`
	MatchRegex    *string               `json:"match_regex"`
	Managed       bool                  `json:"managed" gorm:"-"`
	ManagedSource *ManagedChannelSource `json:"managed_source,omitempty" gorm:"-"`
}

func (c *Channel) UnmarshalJSON(data []byte) error {
	type alias Channel
	aux := struct {
		*alias
		Proxy        *bool   `json:"proxy"`
		ChannelProxy *string `json:"channel_proxy"`
	}{alias: (*alias)(c)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if aux.Proxy != nil {
		c.Proxy = *aux.Proxy
	}
	if aux.ChannelProxy != nil {
		c.ChannelProxy = aux.ChannelProxy
	}
	return nil
}

type ManagedChannelSource struct {
	SiteID          int    `json:"site_id"`
	SiteAccountID   int    `json:"site_account_id"`
	SiteUserGroupID *int   `json:"site_user_group_id,omitempty"`
	GroupKey        string `json:"group_key"`
}

type BaseUrl struct {
	URL   string `json:"url"`
	Delay int    `json:"delay"`
}

type CustomHeader struct {
	HeaderKey   string `json:"header_key"`
	HeaderValue string `json:"header_value"`
}

type ChannelKey struct {
	ID               int     `json:"id" gorm:"primaryKey"`
	ChannelID        int     `json:"channel_id"`
	Enabled          bool    `json:"enabled" gorm:"default:true"`
	ChannelKey       string  `json:"channel_key"`
	StatusCode       int     `json:"status_code"`
	LastUseTimeStamp int64   `json:"last_use_time_stamp"`
	TotalCost        float64 `json:"total_cost"`
	Remark           string  `json:"remark"`
}

type ChannelKeySelectOptions struct {
	ExcludeKeyIDs  map[int]struct{}
	PreferredKeyID int
}

// ChannelUpdateRequest 渠道更新请求 - 仅包含变更的数据
type ChannelUpdateRequest struct {
	ID            int                    `json:"id" binding:"required"`
	Name          *string                `json:"name,omitempty"`
	Type          *outbound.OutboundType `json:"type,omitempty"`
	Enabled       *bool                  `json:"enabled,omitempty"`
	BaseUrls      *[]BaseUrl             `json:"base_urls,omitempty"`
	Model         *string                `json:"model,omitempty"`
	CustomModel   *string                `json:"custom_model,omitempty"`
	ProxyMode     *ProxyUsageMode        `json:"proxy_mode,omitempty"`
	ProxyConfigID *int                   `json:"proxy_config_id,omitempty"`
	Proxy         *bool                  `json:"-"`
	AutoSync        *bool                  `json:"auto_sync,omitempty"`
	SkipHealthProbe *bool                  `json:"skip_health_probe,omitempty"`
	AutoGroup       *AutoGroupType         `json:"auto_group,omitempty"`
	CustomHeader  *[]CustomHeader        `json:"custom_header,omitempty"`
	WSMode        *ChannelWSMode         `json:"ws_mode,omitempty"`
	ChannelProxy  *string                `json:"-"`
	ParamOverride *string                `json:"param_override,omitempty"`
	MatchRegex    *string                `json:"match_regex,omitempty"`

	KeysToAdd    []ChannelKeyAddRequest    `json:"keys_to_add,omitempty"`
	KeysToUpdate []ChannelKeyUpdateRequest `json:"keys_to_update,omitempty"`
	KeysToDelete []int                     `json:"keys_to_delete,omitempty"`

	BypassManagedCheck bool `json:"-"` // 内部使用：允许投影逻辑更新 managed channel
}

type ChannelKeyAddRequest struct {
	Enabled    bool   `json:"enabled"`
	ChannelKey string `json:"channel_key" binding:"required"`
	Remark     string `json:"remark"`
}

type ChannelKeyUpdateRequest struct {
	ID         int     `json:"id" binding:"required"`
	Enabled    *bool   `json:"enabled,omitempty"`
	ChannelKey *string `json:"channel_key,omitempty"`
	Remark     *string `json:"remark,omitempty"`
}

// ChannelFetchModelRequest is used by /channel/fetch-model (not persisted).
type ChannelFetchModelRequest struct {
	Type          outbound.OutboundType `json:"type" binding:"required"`
	BaseURL       string                `json:"base_url" binding:"required"`
	Key           string                `json:"key" binding:"required"`
	ProxyMode     ProxyUsageMode        `json:"proxy_mode"`
	ProxyConfigID *int                  `json:"proxy_config_id"`
}

func (c *Channel) GetBaseUrl() string {
	if c == nil || len(c.BaseUrls) == 0 {
		return ""
	}

	bestURL := ""
	bestDelay := 0
	bestSet := false

	for _, bu := range c.BaseUrls {
		if bu.URL == "" {
			continue
		}
		if !bestSet || bu.Delay < bestDelay {
			bestURL = bu.URL
			bestDelay = bu.Delay
			bestSet = true
		}
	}

	return bestURL
}

func (c *Channel) GetChannelKey(opts ...ChannelKeySelectOptions) ChannelKey {
	if c == nil || len(c.Keys) == 0 {
		return ChannelKey{}
	}

	var selectOpts ChannelKeySelectOptions
	if len(opts) > 0 {
		selectOpts = opts[0]
	}

	if selectOpts.PreferredKeyID > 0 {
		for _, k := range c.Keys {
			if k.ID != selectOpts.PreferredKeyID || !k.Enabled || k.ChannelKey == "" {
				continue
			}
			if _, excluded := selectOpts.ExcludeKeyIDs[k.ID]; excluded {
				break
			}
			return k
		}
	}

	best := ChannelKey{}
	bestCost := 0.0
	bestSet := false

	for _, k := range c.Keys {
		if !k.Enabled || k.ChannelKey == "" {
			continue
		}
		if _, excluded := selectOpts.ExcludeKeyIDs[k.ID]; excluded {
			continue
		}
		if !bestSet || k.TotalCost < bestCost {
			best = k
			bestCost = k.TotalCost
			bestSet = true
		}
	}

	if !bestSet {
		return ChannelKey{}
	}
	return best
}

File: web/src/api/endpoints/channel.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { logger } from '@/lib/logger';
import { formatCount, formatMoney, formatTime } from '@/lib/utils';
import { StatsChannel, type StatsMetricsFormatted } from './stats';
import type { ProxyMode } from './proxy-pool';
/**
 * 渠道类型枚举
 */
export enum ChannelType {
    OpenAIChat = 0,
    OpenAIResponse = 1,
    Anthropic = 2,
    Gemini = 3,
    Volcengine = 4,
    OpenAIEmbedding = 5,
}

/**
 * 自动分组类型枚举
 */
export enum AutoGroupType {
    None = 0,   // 不自动分组
    Fuzzy = 1,  // 模糊匹配
    Exact = 2,  // 准确匹配
    Regex = 3,  // 正则匹配
}

export type ChannelWSMode = 'inherit' | 'off' | 'passthrough' | 'transform';

export type BaseUrl = {
    url: string;
    delay: number;
};

export type CustomHeader = {
    header_key: string;
    header_value: string;
};

export type ChannelKey = {
    id: number;
    channel_id: number;
    enabled: boolean;
    channel_key: string;
    status_code: number;
    last_use_time_stamp: number;
    total_cost: number;
    remark: string;
};

export type ManagedChannelSource = {
    site_id: number;
    site_account_id: number;
    site_user_group_id?: number | null;
    group_key: string;
};

/**
 * 渠道完整数据（与后端 model.Channel 对齐；数组字段在前端保证为 []）
 */
export type Channel = {
    id: number;
    name: string;
    type: ChannelType;
    enabled: boolean;
    base_urls: BaseUrl[];
    keys: ChannelKey[];
    model: string;
    custom_model: string;
    proxy_mode: Exclude<ProxyMode, 'inherit'>;
    proxy_config_id?: number | null;
    auto_sync: boolean;
    /** Skip group health probes for this channel (#102) */
    skip_health_probe?: boolean;
    auto_group: AutoGroupType;
    custom_header: CustomHeader[];
    ws_mode: ChannelWSMode;
    param_override?: string | null;
    match_regex?: string | null;
    managed: boolean;
    managed_source?: ManagedChannelSource | null;
    stats: StatsChannel;
};

// Internal type: backend may return null for slice fields; normalize to [] in select()
type ChannelServer = Omit<Channel, 'base_urls' | 'custom_header' | 'keys'> & {
    base_urls: BaseUrl[] | null;
    custom_header: CustomHeader[] | null;
    keys: ChannelKey[] | null;
};

/**
 * 创建渠道请求：必填字段 + 可选字段
 */
export type CreateChannelRequest = {
    name: string;
    type: ChannelType;
    enabled?: boolean;
    base_urls: BaseUrl[];
    keys: Array<Pick<ChannelKey, 'enabled' | 'channel_key' | 'remark'>>;
    model: string;
    custom_model?: string;
    proxy_mode?: Exclude<ProxyMode, 'inherit'>;
    proxy_config_id?: number | null;
    auto_sync?: boolean;
    skip_health_probe?: boolean;
    auto_group?: AutoGroupType;
    custom_header?: CustomHeader[];
    ws_mode?: ChannelWSMode;
    param_override?: string | null;
    match_regex?: string | null;
};

/**
 * 更新渠道请求：id + 可选字段 + keys diff
 */
export type UpdateChannelRequest = {
    id: number;
    name?: string;
    type?: ChannelType;
    enabled?: boolean;
    base_urls?: BaseUrl[];
    model?: string;
    custom_model?: string;
    proxy_mode?: Exclude<ProxyMode, 'inherit'>;
    proxy_config_id?: number | null;
    auto_sync?: boolean;
    skip_health_probe?: boolean;
    auto_group?: AutoGroupType;
    custom_header?: CustomHeader[];
    ws_mode?: ChannelWSMode;
    param_override?: string | null;
    match_regex?: string | null;
    // keys diff
    keys_to_add?: Array<Pick<ChannelKey, 'enabled' | 'channel_key' | 'remark'>>;
    keys_to_update?: Array<{ id: number; enabled?: boolean; channel_key?: string; remark?: string }>;
    keys_to_delete?: number[];
};

export type FetchModelRequest = {
    type: ChannelType;
    base_urls: BaseUrl[];
    keys: Array<Pick<ChannelKey, 'enabled' | 'channel_key'>>;
    proxy_mode?: Exclude<ProxyMode, 'inherit'>;
    proxy_config_id?: number | null;
    match_regex?: string | null;
    custom_header?: CustomHeader[];
};

/**
 * 获取渠道列表 Hook
 * 
 * @example
 * const { data: channels, isLoading, error } = useChannelList();
 * 
 * if (isLoading) return <Loading />;
 * if (error) return <Error message={error.message} />;
 * 
 * channels?.forEach(channel => console.log(channel.raw.name));
 */

function invalidateChannelRelated(queryClient: ReturnType<typeof useQueryClient>) {
    queryClient.invalidateQueries({ queryKey: ['channels', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['models', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['models', 'channel'] });
    queryClient.invalidateQueries({ queryKey: ['proxy-pool'] });
    queryClient.invalidateQueries({ queryKey: ['groups', 'list'] });
    // Channel config changes reset circuit/sticky + recent fail window on the server.
    queryClient.invalidateQueries({ queryKey: ['runtime'] });
    queryClient.invalidateQueries({ queryKey: ['group-health'] });
    queryClient.invalidateQueries({ queryKey: ['site-channel', 'list'] });
}

export function useChannelList() {
    return useQuery({
        queryKey: ['channels', 'list'],
        queryFn: async () => {
            return apiClient.get<ChannelServer[]>('/api/v1/channel/list');
        },
        select: (data) => data.map((item) => ({
            raw: ({
                ...item,
                managed: item.managed ?? false,
                managed_source: item.managed_source ?? null,
                base_urls: item.base_urls ?? [],
                custom_header: item.custom_header ?? [],
                ws_mode: item.ws_mode ?? 'inherit',
                keys: item.keys ?? [],
                proxy_mode: item.proxy_mode ?? 'direct',
                proxy_config_id: item.proxy_config_id ?? null,
            }) satisfies Channel,
            formatted: {
                input_token: formatCount(item.stats.input_token),
                output_token: formatCount(item.stats.output_token),
                cache_read_token: formatCount(item.stats.cache_read_token || 0),
                cache_write_token: formatCount(item.stats.cache_write_token || 0),
                total_token: formatCount(item.stats.input_token + item.stats.output_token),
                input_cost: formatMoney(item.stats.input_cost),
                output_cost: formatMoney(item.stats.output_cost),
                total_cost: formatMoney(item.stats.input_cost + item.stats.output_cost),
                request_success: formatCount(item.stats.request_success),
                request_failed: formatCount(item.stats.request_failed),
                request_count: formatCount(item.stats.request_success + item.stats.request_failed),
                wait_time: formatTime(item.stats.wait_time),
                cache_hit_rate: (() => {
                    const input = item.stats.input_token || 0;
                    const cacheRead = item.stats.cache_read_token || 0;
                    const denom = Math.max(input, cacheRead);
                    return denom > 0 ? (cacheRead / denom) * 100 : 0;
                })(),
            }
        })) as Array<{ raw: Channel; formatted: StatsMetricsFormatted }>,
        refetchInterval: 30000,
    });
}

/**
 * 创建渠道 Hook
 * 
 * @example
 * const createChannel = useCreateChannel();
 * 
 * createChannel.mutate({
 *   name: 'OpenAI',
 *   type: ChannelType.OpenAIChat,
 *   base_urls: [{ url: 'https://api.openai.com', delay: 0 }],
 *   keys: [{ enabled: true, channel_key: 'sk-xxx' }],
 *   model: 'gpt-4',
 * });
 */
export function useCreateChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (data: CreateChannelRequest) => {
            return apiClient.post<ChannelServer>('/api/v1/channel/create', data);
        },
        onSuccess: (data) => {
            logger.log('渠道创建成功:', data);
            invalidateChannelRelated(queryClient);
        },
        onError: (error) => {
            logger.error('渠道创建失败:', error);
        },
    });
}

/**
 * 更新渠道 Hook
 * 
 * @example
 * const updateChannel = useUpdateChannel();
 * 
 * updateChannel.mutate({
 *   id: 1,
 *   name: 'OpenAI Updated',
 *   type: ChannelType.OpenAIChat,
 *   enabled: true,
 *   base_urls: [{ url: 'https://api.openai.com', delay: 0 }],
 *   keys_to_add: [{ enabled: true, channel_key: 'sk-xxx' }],
 *   model: 'gpt-4-turbo',
 *   proxy: false,
 * });
 */
export function useUpdateChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (data: UpdateChannelRequest) => {
            return apiClient.post<ChannelServer>('/api/v1/channel/update', data);
        },
        onSuccess: (data) => {
            logger.log('渠道更新成功:', data);
            invalidateChannelRelated(queryClient);
        },
        onError: (error) => {
            logger.error('渠道更新失败:', error);
        },
    });
}

/**
 * 删除渠道 Hook
 * 
 * @example
 * const deleteChannel = useDeleteChannel();
 * 
 * deleteChannel.mutate(1); // 删除 ID 为 1 的渠道
 */
export function useDeleteChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (id: number) => {
            return apiClient.delete<null>(`/api/v1/channel/delete/${id}`);
        },
        onSuccess: () => {
            logger.log('渠道删除成功');
            invalidateChannelRelated(queryClient);
        },
        onError: (error) => {
            logger.error('渠道删除失败:', error);
        },
    });
}

/**
 * 启用/禁用渠道 Hook
 * 
 * @example
 * const enableChannel = useEnableChannel();
 * 
 * enableChannel.mutate({ id: 1, enabled: true }); // 启用 ID 为 1 的渠道
 * enableChannel.mutate({ id: 1, enabled: false }); // 禁用 ID 为 1 的渠道
 */
export function useEnableChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (data: { id: number; enabled: boolean }) => {
            return apiClient.post<null>('/api/v1/channel/enable', data);
        },
        onSuccess: () => {
            logger.log('渠道状态更新成功');
            invalidateChannelRelated(queryClient);
        },
        onError: (error) => {
            logger.error('渠道状态更新失败:', error);
        },
    });
}

/**
 * 获取渠道模型列表 Hook
 * 
 * @example
 * const fetchModel = useFetchModel();
 * 
 * fetchModel.mutate({
 *   type: ChannelType.OpenAIChat,
 *   base_urls: [{ url: 'https://api.openai.com', delay: 0 }],
 *   keys: [{ enabled: true, channel_key: 'sk-xxx' }],
 *   proxy: false,
 * });
 * 
 * // 在 onSuccess 中获取模型列表
 * fetchModel.data // ['gpt-4', 'gpt-3.5-turbo', ...]
 */
export function useFetchModel() {
    return useMutation({
        mutationFn: async (data: FetchModelRequest) => {
            return apiClient.post<string[]>('/api/v1/channel/fetch-model', data);
        },
        onSuccess: (data) => {
            logger.log('模型列表获取成功:', data);
        },
        onError: (error) => {
            logger.error('模型列表获取失败:', error);
        },
    });
}

/**
 * 获取渠道最后同步时间 Hook
 * 
 * @example
 * const lastSyncTime = useLastSyncTime();
 * 
 * if (lastSyncTime) {
 *   console.log('最后同步时间:', new Date(lastSyncTime).toLocaleString());
 * }
 */
export function useLastSyncTime() {
    return useQuery({
        queryKey: ['channels', 'last-sync-time'],
        queryFn: async () => {
            return apiClient.get<string>('/api/v1/channel/last-sync-time');
        },
        refetchInterval: 30000,
    });
}
/**
 * 同步渠道 Hook
 * 
 * @example
 * const syncChannel = useSyncChannel();
 * 
 * syncChannel.mutate();
 */
export function useSyncChannel() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async () => {
            return apiClient.post<null>('/api/v1/channel/sync');
        },
        onSuccess: () => {
            logger.log('渠道同步成功');
            queryClient.invalidateQueries({ queryKey: ['channels', 'last-sync-time'] });
        },
        onError: (error) => {
            logger.error('渠道同步失败:', error);
        },
    });
}
